lib/utils/genmap/genmap.go

```diff
@@ -0,0 +1,317 @@
+	ent, isNew := m.ensureEntry(key)
+	if isNew {
+		return
+	}
+
+	select {
+	case ent.wantRegen <- struct{}{}:
+	default:
+	}
+}
+
+// Terminate halts generation for the target key. Note that future calls to Get/Generate will restart generation. For this
+// reason a robust 'cleanup' strategy should also include providing a generator that returns errors when called for keys
+// that are no longer needed. This will ensure that concurrent get/generate calls can't result in dangling entries long-term,
+// since the entry will eventually get cleaned up naturally when it hits MaxFailures.
+func (m *GenMap[K, V]) Terminate(key K) {
+	m.rw.RLock()
+	ent, ok := m.entries[key]
+	m.rw.RUnlock()
+	if !ok {
+		return
+	}
+
+	m.rw.Lock()
+	ent, ok = m.entries[key]
+	if ok {
+		delete(m.entries, key)
+	}
+	m.rw.Unlock()
+	if !ok {
+		return
+	}
+	ent.cancel()
+}
+
+// Get loads the value associated with the target key. If the key was not yet being generated by this
+// map instance, generation will be started. This method is non-blocking, except in the case where
+// the key being accessed is/was just added to the mapping.
+func (m *GenMap[K, V]) Get(ctx context.Context, key K) (val V, err error) {
+	ent, _ := m.ensureEntry(key)
+
+	// optimistically load the value pointer (this will always succeed unless the genmap just
+	// started tracking this value, so over time most calls should follow this path).
+	if vp := ent.ptr.Load(); vp != nil {
+		return vp.read()
+	}
+
+	// note that we aren't selecting on the per-entry context. each entry always generates at
+	// lease one value before exiting, so we only care about exiting early if the map as a whole
+	// is closed.
+	select {
+	case <-ent.init:
+		return ent.ptr.Load().read()
+	case <-m.ctx.Done():
+		return val, ErrGenMapClosed
+	case <-ctx.Done():
+		return val, ctx.Err()
+	}
+}
+
+// ensureEntry gets the entry for the target key if it exists, falling back to setting up a new one.
+func (m *GenMap[K, V]) ensureEntry(key K) (ent *entry[V], isNew bool) {
+	m.rw.RLock()
+	ent, ok := m.entries[key]
+	m.rw.RUnlock()
+	if ok {
+		return ent, false
+	}
+
+	// create an entry-level context that is a child of the
+	// main close context.
+	ctx, cancel := context.WithCancel(m.ctx)
+
+	newEnt := &entry[V]{
+		init:      make(chan struct{}),
+		wantRegen: make(chan struct{}, 1),
+		ctx:       ctx,
+		cancel:    cancel,
+	}
+
+	m.rw.Lock()
+	// check if another entry was set up concurrently
+	ent, ok = m.entries[key]
+	if !ok {
+		m.entries[key] = newEnt
+	}
+	m.rw.Unlock()
+	if ok {
+		cancel()
+		return ent, false
+	}
+
+	// start background opts for new entry
+	go m.generate(key, newEnt)
+
+	return newEnt, true
+}
+
+// generate is the background process that generates/regenerates the value associated with a given key.
+func (m *GenMap[K, V]) generate(key K, entry *entry[V]) {
+	defer func() {
+		m.rw.RLock()
+		current := m.entries[key]
+		m.rw.RUnlock()
+		if current != entry {
+			return
+		}
```

Comment:
In what case this is true, besides when `Terminate` is called?

Reply:
`Terminate` should be the only case currently where the entry was removed by a different goroutine.  For both the `Close` case and the too many errors case, this goroutine should be the one unregistering itself.  In theory, we could set up some variables that let us distinguish between remote unregister and local unregister, but IMO it's not worth the effort.  Deciding based on the pointer value is more reliable and future-proof I think.

Reply:
I was just thinking since `Terminate` would be rare, the logic would be simpler, to not do the `m.rw.RLock()` block, and just go straight to `m.rw.Lock()` block. good to have it too for performance.